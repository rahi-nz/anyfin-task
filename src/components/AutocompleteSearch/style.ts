import styled from 'styled-components';

export const Container = styled.section`
  text-align: center;
`;

export const List = styled.ul`
  text-align: center;
  max-height: 300px;
  overflow-y: auto;
  background: #005993;
  box-shadow: 0 4px 6px rgb(32 33 36 / 28%);
  border-radius: 20px;
  padding: 10px 30px;
`;

export const Item = styled.li.attrs(() => ({
  'data-test-id': 'suggested',
}))`
  text-align: left;
  padding: 10px 0;
  a {
    color: #fff;
    display: block;
  }
`;
