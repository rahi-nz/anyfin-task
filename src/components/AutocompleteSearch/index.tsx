import { Container, List } from './style';
import React, { useCallback } from 'react';
import { debounce } from 'lodash';
import Link from 'next/link';
import { TextInput, usePersistedState } from 'components';
import { Country } from 'types/countries';
import { fields } from 'config';

const AutocompleteSearch = () => {
  const [countries, setCountries] = usePersistedState<Country[]>('countries', []);
  /**
   * debounce: to fire it only when the user has finished typing
   * useCallback: store the reference
   */
  const debounceLoadData = useCallback(
    debounce(async query => {
      if (query.length > 0) {
        const response = await fetch(
          `https://restcountries.eu/rest/v2/name/${query}?fields=${fields.join(';')}`,
          {
            method: 'get',
          },
        ).then(res => res.json());
        setCountries(response);
      } else {
        setCountries([]);
      }
    }, 1000),
    [],
  );
  const onInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    debounceLoadData(e.target.value);
  };

  return (
    <Container>
      <TextInput
        name="country"
        label="Search country by name"
        placeholder="Search country by name"
        search
        onChange={onInputChange}
      />
      {countries.length > 0 && (
        <List>
          {countries.map(country => (
            <li key={country.name}>
              <Link href={`/country/${country.alpha3Code}`}>
                <a>{country.name}</a>
              </Link>
            </li>
          ))}
        </List>
      )}
    </Container>
  );
};

export default AutocompleteSearch;
