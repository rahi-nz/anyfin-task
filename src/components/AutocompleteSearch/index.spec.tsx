/**
 * @jest-environment jsdom
 */

import renderer, { act } from 'react-test-renderer';
import { waitFor } from '@testing-library/react';
import fetchMock from 'jest-fetch-mock';
import AutocompleteSearch from './index';

describe('<AutocompleteSearch/>', () => {
  let rootElement;
  let testRenderer;
  const event = {
    target: { value: 'Germany' },
  };
  beforeEach(() => {
    // eslint-disable-next-line prefer-const,react/display-name
    rootElement = () => <AutocompleteSearch />;
    // eslint-disable-next-line prefer-const
    testRenderer = renderer.create(rootElement());
    fetchMock.resetMocks();
  });

  it('should call fetchData after 1000ms after typing', async () => {
    fetchMock.mockResponseOnce(
      JSON.stringify([
        {
          currencies: [{ code: 'EUR', name: 'Euro', symbol: 'â‚¬' }],
          name: 'Germany',
          alpha3Code: 'DEU',
          capital: 'Berlin',
          population: 81770900,
        },
      ]),
    );
    await act(async () => {
      testRenderer.root.findByType('input').props.onChange(event);
      await waitFor(
        () => {
          expect(
            testRenderer.root.findAllByProps({ 'data-test-id': 'suggested' }).length,
          ).toBe(1);
        },
        {
          timeout: 4000,
        },
      );
    });
  });

  it('should not call fetchData before 1000ms after typing', async () => {
    await act(async () => {
      testRenderer.root.findByType('input').props.onChange(event);
      await waitFor(
        () => {
          expect(
            testRenderer.root.findAllByProps({ 'data-test-id': 'suggested' }).length,
          ).toBe(0);
        },
        {
          timeout: 500,
        },
      );
    });
  });

  it('Should render correctly', () => {
    expect(testRenderer.toJSON()).toMatchSnapshot();
  });
});
