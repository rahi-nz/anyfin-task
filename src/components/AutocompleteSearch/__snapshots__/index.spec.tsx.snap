// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AutocompleteSearch/> Should render correctly 1`] = `
<section
  className="style__Container-sc-bmdzv0-0 hJhEuo"
>
  <div
    className="style__Container-sc-16aipdw-0 dWJwII"
  >
    <label
      htmlFor="country"
    >
      Search country by name
    </label>
    <div
      className="style__InputContainer-sc-16aipdw-1 bWdXEG"
    >
      <div
        style={
          Object {
            "boxSizing": "border-box",
            "display": "inline-block",
            "margin": 0,
            "maxWidth": "100%",
            "overflow": "hidden",
            "position": "relative",
          }
        }
      >
        <div
          style={
            Object {
              "boxSizing": "border-box",
              "display": "block",
              "maxWidth": "100%",
            }
          }
        >
          <img
            alt=""
            aria-hidden={true}
            role="presentation"
            src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMjAiIGhlaWdodD0iMjAiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyIgdmVyc2lvbj0iMS4xIi8+"
            style={
              Object {
                "border": "none",
                "display": "block",
                "margin": 0,
                "maxWidth": "100%",
                "padding": 0,
              }
            }
          />
        </div>
        <noscript>
          <img
            alt="anyfin"
            decoding="async"
            src="/_next/image?url=%2Ficons%2Fsearch.svg&w=48&q=75"
            srcSet="/_next/image?url=%2Ficons%2Fsearch.svg&w=32&q=75 1x, /_next/image?url=%2Ficons%2Fsearch.svg&w=48&q=75 2x"
            style={
              Object {
                "border": "none",
                "bottom": 0,
                "boxSizing": "border-box",
                "display": "block",
                "height": 0,
                "left": 0,
                "margin": "auto",
                "maxHeight": "100%",
                "maxWidth": "100%",
                "minHeight": "100%",
                "minWidth": "100%",
                "objectFit": undefined,
                "objectPosition": undefined,
                "padding": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "width": 0,
              }
            }
          />
        </noscript>
        <img
          alt="anyfin"
          decoding="async"
          src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
          style={
            Object {
              "border": "none",
              "bottom": 0,
              "boxSizing": "border-box",
              "display": "block",
              "height": 0,
              "left": 0,
              "margin": "auto",
              "maxHeight": "100%",
              "maxWidth": "100%",
              "minHeight": "100%",
              "minWidth": "100%",
              "objectFit": undefined,
              "objectPosition": undefined,
              "padding": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "width": 0,
            }
          }
        />
      </div>
      <input
        className="style__Input-sc-16aipdw-2 dJIGoW"
        id="country"
        onChange={[Function]}
        placeholder="Search country by name"
        type="text"
      />
    </div>
  </div>
</section>
`;
